### PR TITLE
[jvm-package] jvm test should clean up after themselfs

### PR DESCRIPTION
**Pull request by trams**
I noticed that some tests do not fully clean up their tmp files so I fixed that.
Nothing else

### CODE REVIEW

**Comment by CodingCat** [comment]
which test doesn't clean? the PR looks like extracting temp dir creation to a class instead of adding anything?

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PersistenceSuite.scala [hunk 1]
```diff
@@ -124,6 +98,8 @@ class PersistenceSuite extends FunSuite with PerTest with BeforeAndAfterAll {
   }
 
   test("test persistence of MLlib pipeline with XGBoostClassificationModel") {
+    val tempDir = createTmpFolder("MLLib").toFile
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PersistenceSuite.scala]: why we need to create tempDir in each test manually?

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/TmpFolderPerSuite.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package ml.dmlc.xgboost4j.scala.spark
+
+import java.nio.file.{Files, Path}
+
+import org.apache.spark.network.util.JavaUtils
+import org.scalatest.{BeforeAndAfterAll, FunSuite}
+
+trait TmpFolderPerSuite extends BeforeAndAfterAll { self: FunSuite =>
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/TmpFolderPerSuite.scala]: what's the issue to put it in PerTest?

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PersistenceSuite.scala (hunk 1 above)
Comment (trams) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PersistenceSuite.scala]: As far as I know in scalatest different tests are not isolated.
I create a temp folder once per suite. So one need to create temp folders in each test explicitly to make them isolated.

Alternative solution is to change a trait to create a temp folder per test

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/TmpFolderPerSuite.scala (hunk 2 above)
Comment (trams) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/TmpFolderPerSuite.scala]: I wanted to have an extra flexibility if we want to have a test which create temp files but we do not need SparkSession

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PersistenceSuite.scala (hunk 1 above)
Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PersistenceSuite.scala]: but what's the problem with reusing the created dir here?

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/CheckpointManagerSuite.scala [hunk 3]
```diff
@@ -17,12 +17,11 @@
...
 
 import java.io.File
-import java.nio.file.Files
 
-import org.scalatest.{BeforeAndAfterAll, FunSuite}
+import org.scalatest.FunSuite
 import org.apache.hadoop.fs.{FileSystem, Path}
 
-class CheckpointManagerSuite extends FunSuite with PerTest with BeforeAndAfterAll {
+class CheckpointManagerSuite extends FunSuite with TmpFolderPerSuite with PerTest {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/CheckpointManagerSuite.scala]: I may miss something, why this change can address the issue?

is it just moving the beforeAll&afterAll to somewhere else?

Comment (trams) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/CheckpointManagerSuite.scala]: Before this change in CheckpointManagerSuite each test created the temp directory in /tmp (on linux) and never removed it.

(Note that PerTest neither did not nor does not now create any temp directories. It just create a SparkSession to use)

After this change CheckpointManagerSuite will include TmpFolderPerSuite .
TmpFolderPerSuite  will create a temp directory and then it will be removed.
Also after this change all temp folders are created as children of this root temp folder (see line 35 in this file)

Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/CheckpointManagerSuite.scala]: oh. yes, we manually created "test" dir there....

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PersistenceSuite.scala (hunk 1 above)
Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PersistenceSuite.scala]: I would prefer keeping writing test easy by forcing devs to remember call createTmpFolder manually,

if reusing the directory doesn't bring any issue, can we just reuse the created one in beforeAll()?

Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/PersistenceSuite.scala]: @trams would you please address this so that we can merge it?

**Comment by trams** [comment]
Yes, I will address it

**Comment by trams** [comment]
@CodingCat , I updated review as you asked. I also rebased on a fresh master and CI gives a green light.

**Comment by CodingCat** [comment]
thanks, merged!
